### PR TITLE
[Consensus] Switch RainTree broadcast to star-pattern broadcast

### PR DESCRIPTION
**Pull request by Olshansk**
## Description

tl;dr Stabilize LocalNet

After #374 was merged, LocalNet was not as stable as before. One of the reasons for this is because we were using `P2P.Broadcast()`, which uses `RainTree` by default, whereas `HotPOKT` expects a star-like pattern.

Followup work / documentation may involve exposing or documenting the difference between the following:
- `BroadcastToValidators` - Direct O(N) send from a `leader` to all `validator` replicas used by `HotPOKT`
- `BroadcastToStakedActors` - O(log3N) broadcast using `RainTree` from an originator to all staked nodes
- `RandomBroadcast` - Random gossip to all nodes (staked or not), including full nodes

## Issue

Bugfix

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [x] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Rename `sendToNode` to `sendToLeader`
- Rename `broadcastToNodes` to `broadcastToValidators`
- Change HotPOKT broadcast from `RainTree` to `start pattern`

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

**Comment by Olshansk** [comment]
@gokutheengineer FYI that I'm also working on https://github.com/pokt-network/pocket/pull/404/files in parallel. It's just a draft that refactors some stuff. I'll aim to send it your way for review by monday.

File: consensus/helpers.go [hunk 1]
```diff
@@ -137,35 +137,41 @@ func protoHash(m proto.Message) string {
...
 		return
 	}
-	if err := m.GetBus().GetP2PModule().Broadcast(anyConsensusMessage); err != nil {
-		m.nodeLogError(typesCons.ErrBroadcastMessage.Error(), err)
-		return
+
+	for _, val := range m.validatorMap {
+		if err := m.GetBus().GetP2PModule().Send(cryptoPocket.AddressFromString(val.GetAddress()), anyConsensusMessage); err != nil {
+			m.nodeLogError(typesCons.ErrBroadcastMessage.Error(), err)
+		}
```
Comment (deblasis) [inline consensus/helpers.go:174]: I am not sure I like this... we have P2P concerns leaking into the consensus module. Do we want that?

Comment (deblasis) [reply consensus/helpers.go:174]: Also https://github.com/pokt-network/pocket-network-protocol/blob/main/p2p/README.md#i-requirements

image

If this is a `HACK` I guess it should be written there otherwise the reader could get very confused very quickly.

File: consensus/helpers.go [hunk 2]
```diff
@@ -137,35 +137,41 @@ func protoHash(m proto.Message) string {
...
 	}
+
 	if err := m.GetBus().GetP2PModule().Send(cryptoPocket.AddressFromString(m.idToValAddrMap[*m.leaderId]), anyConsensusMessage); err != nil {
 		m.nodeLogError(typesCons.ErrSendMessage.Error(), err)
 		return
 	}
 }
 
-func (m *consensusModule) broadcastToNodes(msg *typesCons.HotstuffMessage) {
+// A start-pattern broadcast (i.e. leader to all other validators) broadcast
```
Comment (deblasis) [inline consensus/helpers.go:161]: star-pattern?

**Review by deblasis — CHANGES_REQUESTED** [review]
In the title:

> "Start-Pattern"

I guess you mean "star-pattern" 

> After https://github.com/pokt-network/pocket/pull/374 was merged, LocalNet was not as stable as before.

Care to comment specifically about the instability you are referring to? Is there anything that I have missed apart from #401 fixed by #400? 

If you are referring to the fact that previously we were "broadcasting" to a static list of validators and now we are instead considering all staked actors, it sounds more like tech debt. I just want to make sure I am in sync.

What is this PR fixing?

File: consensus/helpers.go (hunk 1 above)
Comment (Olshansk) [reply consensus/helpers.go:174]: It's not a hack, but rather a previous incorrect implementation. See the details from the github message:

* `BroadcastToValidators` - Direct `O(N)` send from a leader to all validator replicas used by HotPOKT
* `BroadcastToStakedActors` - `O(log3N)` broadcast using RainTree from an originator to all staked nodes
* `RandomBroadcast` - Random gossip to all nodes (staked or not), including full nodes

[This article](https://medium.com/ontologynetwork/hotstuff-the-consensus-protocol-behind-facebooks-librabft-a5503680b151) has a very nice visual for basic hotstuff:

Screenshot 2022-12-16 at 2 04 39 PM

- During consensus: the validators use star pattern
- During tx & block propagation: staked actors use structured gossip
- During state sync and block propagation with full nodes: random gossip

Does that make sense?

**Comment by Olshansk** [comment]
> I guess you mean "star-pattern"

Correct.

> Care to comment specifically about the instability you are referring to? Is there anything that I have missed apart from https://github.com/pokt-network/pocket/issues/401 fixed by https://github.com/pokt-network/pocket/pull/400?

I created a couple videos since I think it'll be easier to describe rather than writing it out. The tl;dr is "Less consistent" probably due to networking duplication/stability issues but I'm not 100% sure:

- Before 400: https://drive.google.com/file/d/1xlT6O8AlYfKGBMl-ED8HHtXRc7k_gj5K/view?usp=share_link
- After 400: https://drive.google.com/file/d/14RwupsxJRwVexpwYuZT2XWxjWEXoSOdA/view?usp=share_link

> If you are referring to the fact that previously we were "broadcasting" to a static list of validators and now we are instead considering all staked actors, it sounds more like tech debt. I just want to make sure I am in sync.

The consensus module, during consensus, should not be doing RainTree broadcast to all staked actors. It should be doing a direct send to only staked validators. Going to copy-paste my answer from the comment [here](https://github.com/pokt-network/pocket/pull/405#discussion_r1051198643) as well.

* `BroadcastToValidators` - Direct `O(N)` send from a leader to all validator replicas used by HotPOKT
* `BroadcastToStakedActors` - `O(log3N)` broadcast using RainTree from an originator to all staked nodes
* `RandomBroadcast` - Random gossip to all nodes (staked or not), including full nodes

[This article](https://medium.com/ontologynetwork/hotstuff-the-consensus-protocol-behind-facebooks-librabft-a5503680b151) has a very nice visual for basic hotstuff:

Screenshot 2022-12-16 at 2 04 39 PM

cc @gokutheengineer since its good context

**Review by deblasis — APPROVED** [review]

**Comment by Olshansk** [comment]
@deblasis I wanted to follow up on why I haven't merged this in yet.

I'm still driving for continuing to formalize [the following](https://github.com/pokt-network/pocket/pull/405#issuecomment-1355722138) as our main communication patterns:

Screenshot 2022-12-22 at 11 34 06 AM

But then I came by [this tweet](https://twitter.com/ittaia/status/1605141881872056320) from Ittai [1]:

Screenshot 2022-12-22 at 11 36 07 AM

Which made me think that REGARDLESS of whether we are using the `⭐ -pattern` or `🌧️ 🌲` during consensus, this should still work. The issue @gokutheengineer brought up related to `block is nil` may be hiding an underlying issue that I'm investigating in #404, that I want to figure out first.

[1] Must subscribe to blogpost if you're not familiar with who Ittai is: https://decentralizedthoughts.github.io/about-ittai/

**Comment by deblasis** [comment]
@Olshansk:

I think Ittai is one of the reasons I am here. Reading some of his work allowed me to do an OK job during my interview with Andrew :)

Regarding the investigation, not sure if related but there's a configuration value, the one used for the mempool deduping, that's not there in the config. Just a hunch. 

https://github.com/pokt-network/pocket/pull/403/files#diff-aea3d8b43c12c0ea281bee2029f445f9aeecd65e5280383a6d6bb14d88562992R29
